### PR TITLE
bpo-28556: Don't simplify unions at runtime

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/typing/issues/536

cc @gvanrossum 

https://bugs.python.org/issue28556

### CODE REVIEW

**Review by grimreaper — COMMENTED** [review]
The change itself looks find to me, but its not clear that changing this is the right thing to do. As a general note we're removing old explicit logic and tests rather than just fixing a bug. I didn't notice any discussion of the issue on the mailing list or otherwise. Feel free to direct me if I missed it.

**Comment by ilevkivskyi** [comment]
> The change itself looks find to me, but its not clear that changing this is the right thing to do. As a general note we're removing old explicit logic and tests rather than just fixing a bug. I didn't notice any discussion of the issue on the mailing list or otherwise. Feel free to direct me if I missed it.

Some people see this as a bug, see example in the issue linked above. Also this code was a bug magnet because it needs to use `issubclass()` which is prohibited for various special types.

**Comment by ilevkivskyi** [comment]
@gvanrossum Sorry for being annoying, but if we want to get this in 3.7 we only have time until Monday. Could you please briefly take a look at this PR?

**Comment by gvanrossum** [comment]
No worries. I'll try to review later today.

File: Lib/test/test_typing.py [hunk 1]
```diff
@@ -253,10 +253,9 @@ def test_union_any(self):
     def test_union_object(self):
         u = Union[object]
         self.assertEqual(u, object)
-        u = Union[int, object]
-        self.assertEqual(u, object)
-        u = Union[object, int]
-        self.assertEqual(u, object)
+        u1 = Union[int, object]
+        u2 = Union[object, int]
+        self.assertEqual(u1, u2)
```
Comment (gvanrossum) [inline Lib/test/test_typing.py:258]: Maybe also assert that it isn't `object`?

File: Lib/test/test_typing.py [hunk 2]
```diff
@@ -267,13 +266,9 @@ def test_single_class_disappears(self):
...
 
-    def test_base_class_disappears(self):
-        u = Union[Employee, Manager, int]
-        self.assertEqual(u, Union[int, Employee])
-        u = Union[Manager, int, Employee]
-        self.assertEqual(u, Union[int, Employee])
+    def test_base_class_kept(self):
         u = Union[Employee, Manager]
-        self.assertIs(u, Employee)
+        self.assertIsNot(u, Employee)
```
Comment (gvanrossum) [inline Lib/test/test_typing.py]: Can you check that Employee and Manager are both members of the union?

File: Lib/test/test_typing.py [hunk 3]
```diff
@@ -317,7 +312,6 @@ def test_cannot_instantiate(self):
     def test_union_generalization(self):
         self.assertFalse(Union[str, typing.Iterable[int]] == str)
         self.assertFalse(Union[str, typing.Iterable[int]] == typing.Iterable[int])
-        self.assertTrue(Union[str, typing.Iterable] == typing.Iterable)
```
Comment (gvanrossum) [inline Lib/test/test_typing.py:320]: How about a more positive test of the expected members?

File: Lib/typing.py [hunk 4]
```diff
@@ -203,7 +203,7 @@ def _check_generic(cls, parameters):
 
 def _remove_dups_flatten(parameters):
     """An internal helper for Union creation and substitution: flatten Union's
```
Comment (gvanrossum) [inline Lib/typing.py]: Union's -> Unions

**Review by gvanrossum — COMMENTED** [review]
A few nits about tests; the substance of the code LGTM. (Nice that it was such a clear separate bit of code. :-)

**Comment by ilevkivskyi** [comment]
@gvanrossum Thanks for review! I addressed all comments, plus I think there is one important point to add: this change breaks semantics of `__eq__`, now `Union[Manager, Employee] != Employee`. I can tweak `__eq__`, but the problem is that generics are cashed, so after once writing `List[Employee]`, then writing `List[Union[Manager, Employee]]` will return `List[Employee]` thus again effectively simplifying a union.

**Comment by gvanrossum** [comment]
> this change breaks semantics of __eq__, now Union[Manager, Employee] != Employee. I can tweak __eq__, but the problem is that generics are cashed, so after once writing List[Employee], then writing List[Union[Manager, Employee]] will return List[Employee] thus again effectively simplifying a union.

Sorry, the tenses are confusing. Is the problem with List[Union[Manager, Employee]] returning List[Employee] happening with this version of the PR or would that happen if you "tweaked" `__eq__`?

Honestly given that we don't simplify unions I think it's right that Union[Manager, Employee] != Employee.

**Comment by gvanrossum** [comment]
Also I think this requires a news item.

**Comment by ilevkivskyi** [comment]
I will close and re-open to trigger AppVeyor.

**Comment by ilevkivskyi** [comment]
Restarting AppVeyor again...

**Review by gvanrossum — APPROVED** [review]
You have to merge this manually because the VSTS tests always fail (they are non-blocking, but Miss Islington doesn't know).

**Comment by ilevkivskyi** [comment]
> You have to merge this manually because the VSTS tests always fail (they are non-blocking, but Miss Islington doesn't know).

OK, I see. Will do this now.

**Pull request merged**